### PR TITLE
heater: Add a thermal protection

### PR DESCRIPTION
**Pull request by cruwaller**
Checks the faulty thermistor and shutdown the sytems if
        temperatu is not increased / maintained correctly.

Signed-off-by: cruwaller <cruwaller@gmail.com>

### CODE REVIEW

File: klippy/heater.py [hunk 1]
```diff
@@ -141,6 +141,92 @@ def __init__(self, printer, config):
...
+                    gcode = self.printer.lookup_object('gcode')
+                    gcode.respond_error(errorstr)
+                    self.printer.request_exit('shutdown')
+            self.protection_last_temp = current_temp
+            next_time = self.protection_period_heat
+        logging.debug("check_heating(eventtime {}, next {}) {} / {}".
+                      format(eventtime, (eventtime + next_time),
+                             current_temp, target_temp))
+        return eventtime + next_time
+    def get_min_extrude_status(self):
```
Comment (cruwaller) [inline klippy/heater.py]: Oh, these get_min_extrude_status() and set_min_extrude_temp() need to be removed. 
I will fix this...

**Comment by KevinOConnor** [comment]
Thanks.  I'd happily have pulled this a few weeks ago.  However, I'm pretty close to having a "first order plus delay time" model working.  I think that model would be useful for confirming the heater and thermistor are working properly.  (And, it makes it possible to use good defaults for the pid parameters.)

**Comment by cruwaller** [comment]
ok, no problem :)
It is good to hear that Klipper will have a thermal protection soon. This is very important feature.

**Comment by KevinOConnor** [comment]
Hi.  I'm struggling with implementing checking based on the FOPDT model.  (I can generate the model with great accuracy, but using the model itself to verify the extruder isn't working great as extrusion speed and fans significantly alter the behavior.)  I'm thinking that a simple hysteresis check is a better option.

Are you okay if I take this patch and rework it into an independent module?  I'm thinking of moving the code into extras/verify_heater.py and then changing heater.py to do printer.try_load_module("verify_heater %s" % (self.name,)) so that it is automatically loaded.  Or, would you prefer to resubmit?

-Kevin

**Comment by cruwaller** [comment]
Hi, you are freely to use this code as you want :) 
Main thing is to get thermal protection to Klipper asap. It is the one of the main safety features!

You should add a additional check for ADC values; check e.g. 5 last received values and those should not be same, this eliminates possible ADC sampling errors. Results should always vary a bit...

thanks!

**Comment by KevinOConnor** [comment]
FYI, I decided to implement some simple checks - commit 5208fc38.

**Pull request closed without merging**